nan_json.h

```diff
@@ -0,0 +1,138 @@
+/*********************************************************************
+ * NAN - Native Abstractions for Node.js
+ *
+ * Copyright (c) 2017 NAN contributors
+ *
+ * MIT License <https://github.com/nodejs/nan/blob/master/LICENSE.md>
+ ********************************************************************/
+
+#ifndef NAN_JSON_H_
+#define NAN_JSON_H_
+
+#if ((NODE_MAJOR_VERSION == 0) && (NODE_MINOR_VERSION < 12))
+#define NAN_JSON_H_NEED_PARSE 1
+#else
+#define NAN_JSON_H_NEED_PARSE 0
+#endif
+
+#if (NODE_MAJOR_VERSION >= 7)
+#define NAN_JSON_H_NEED_STRINGIFY 0
+#else
+#define NAN_JSON_H_NEED_STRINGIFY 1
+#endif
+
+class JSON {
+ public:
+  static inline
+  Nan::MaybeLocal<v8::Value> Parse(v8::Local<v8::String> jsonString) {
+#if NAN_JSON_H_NEED_PARSE
+    Nan::HandleScope scope;
+    return instance().parse(jsonString);
+#else
+#if (NODE_MAJOR_VERSION >= 7)
+    return v8::JSON::Parse(Nan::GetCurrentContext(), jsonString);
+#else
+    return v8::JSON::Parse(jsonString);
+#endif
+#endif
+  }
+
+  static inline
+  Nan::MaybeLocal<v8::String> Stringify(v8::Local<v8::Object> jsonObject) {
+#if NAN_JSON_H_NEED_STRINGIFY
+    Nan::HandleScope scope;
+    return instance().stringify(jsonObject)->ToString();
+#else
+    return v8::JSON::Stringify(Nan::GetCurrentContext(), jsonObject);
+#endif
+  }
+
+  static inline
+  Nan::MaybeLocal<v8::String> Stringify(v8::Local<v8::Object> jsonObject,
+    v8::Local<v8::String> gap) {
+#if NAN_JSON_H_NEED_STRINGIFY
+    Nan::HandleScope scope;
+    return instance().stringify(jsonObject, gap)->ToString();
+#else
+    return v8::JSON::Stringify(Nan::GetCurrentContext(), jsonObject, gap);
+#endif
+  }
+
+ private:
+  NAN_DISALLOW_ASSIGN_COPY_MOVE(JSON)
+#if NAN_JSON_H_NEED_PARSE
+  Nan::Callback m_cb_parse;
+#endif
+#if NAN_JSON_H_NEED_STRINGIFY
+  Nan::Callback m_cb_stringify;
+#endif
+
+#if (NAN_JSON_H_NEED_PARSE + NAN_JSON_H_NEED_STRINGIFY)
+  static JSON& instance() {
+    static JSON i;
```

Comment:
Thanks for the review, @bnoordhuis .   The singleton aspect is part of what makes this solution so cool, in my opinion, but you might be right.   Another option would be to ditch the singleton, and have have the `parse` and `stringify` entry points each call into something like this:

```c++
  v8::Local<v8::Value> call(const char *method, v8::Local<v8::Value> arg)
  {
    v8::Local<v8::Object> json = Nan::New(m_persistent);

    v8::Local<v8::Value> thisMethod = json->Get(Nan::New(method).ToLocalChecked());

    if (thisMethod.IsEmpty() || !thisMethod->IsFunction()) {
      return Nan::Undefined();
    }

    v8::Local<v8::Function> methodFunction = v8::Local<v8::Function>::Cast(thisMethod);

    return methodFunction->Call(json, 1, &arg);
  }
```

... changing the entry points to look more like this:

```c++
  static inline v8::Local<v8::Value> Parse(v8::Local<v8::Value> jsonString)
  {
    return JSON::instance().call("parse", jsonString);
  }

  static inline v8::Local<v8::Value> Stringify(v8::Local<v8::Value> jsonObject)
  {
    return JSON::instance().call("stringify", jsonObject);
  }
```

What do you think?